realtimenet/controller.py

```diff
@@ -0,0 +1,131 @@
+from typing import List
+from typing import Union
+
+from realtimenet.camera import VideoSource
+from realtimenet.camera import VideoStream
+from realtimenet.display import DisplayResults
+from realtimenet.engine import InferenceEngine
+from realtimenet.downstream_tasks.postprocess import PostProcessor
+
+import cv2
+import numpy as np
+import torch.nn as nn
+
+
+class Controller:
+
+    def __init__(
+            self,
+            neural_network: nn.Module,
+            post_processors: Union[PostProcessor, List[PostProcessor]],
+            results_display: DisplayResults,
+            camera_id: int,
+            path_in: str = None,
+            path_out: str = None,
+            use_gpu: bool = True):
+
+        video_source = VideoSource(
+            camera_id=camera_id,
+            size=self.inference_engine.expected_frame_size,
+            filename=path_in
+        )
+        self.video_stream = VideoStream(video_source, self.inference_engine.fps)
+        self.inference_engine = InferenceEngine(neural_network, use_gpu=use_gpu)
+        self.postprocessors = post_processors
+        self.results_display = results_display
+        self.path_out = path_out
+        self.video_recorder = None  # created in `display_prediction`
+        self.video_recorder_raw = None  # created in `display_prediction`
+
+    def run_inference(self):
+        clip = np.random.randn(1, self.inference_engine.step_size, self.inference_engine.expected_frame_size[0],
+                               self.inference_engine.expected_frame_size[1], 3)
+
+        frame_index = 0
+        runtime_error = None
+
+        self._start_inference()
+
+        while True:
+            try:
+                frame_index += 1
+
+                # Grab frame if possible
+                img_tuple = self.video_stream.get_image()
+                # If not possible, stop
+                if img_tuple is None:
+                    break
+
+                # Unpack
+                img, numpy_img = img_tuple
+
+                clip = np.roll(clip, -1, 1)
+                clip[:, -1, :, :, :] = numpy_img
+
+                if frame_index == self.inference_engine.step_size:
+                    # A new clip is ready
+                    self.inference_engine.put_nowait(clip)
+
+                frame_index = frame_index % self.inference_engine.step_size
+
+                # Get predictions
+                prediction = self.inference_engine.get_nowait()
+
+                prediction_postprocessed = self.postprocess_prediction(prediction)
+
+                self.display_prediction(img, prediction_postprocessed)
+
+            except Exception as runtime_error:
+                break
```

Comment:
The entire point of writing this controller war to continuously obtain predictions in a loop, which could then be used for some task. I think we should add to option to specify a callback in the constructor, which is called on every loop with the new predictions as an argument. Without further modification, this class is only useful for displaying results.

I am also not entirely happy with the structure of this block, there is a lot going on. Maybe we can refactor `frame_index` and `clip` into attributes which are initialized during `_start_inference`. We could then encapsulate most of the loop into something like `_get_predictions`. I think that would greatly increase readability.